### PR TITLE
Fix: UPDATE Application command &  AutoSync for UPDATE

### PR DESCRIPTION
**Pull request by Debanitrkl**
Fixed the new sub-command that takes in the YAML to create GitOps Application resource on Harness platform.

Sample YAML:

```
gitops:
  name: gitops-application
  projectIdentifier: default_project
  orgIdentifier: default
  type: application
  application:
    metadata:
      namespace: default
      clusterName: gitopscluster
      labels:
        harness.io/serviceRef: gitopsservice
        harness.io/envRef: gitopsenv
    spec:
      source:
        repoURL: https://github.com/harness-community/harnesscd-example-apps
        path: guestbook
        targetRevision: master
      destination:
        server: https://kubernetes.default.svc
        namespace: default
  agentIdentifier: AGENT_NAME
  clusterIdentifier: gitopscluster
  repoIdentifier: gitopsrepo
 ```
 Here, `AGENT_NAME` is a constant and asks for input from the user.
 
 Command:
 ```
 harness gitops-application --file applications.yml apply
 ```
 
 Asks for AGENT_NAME input from the user:
 
 ```
 Trying to create or update gitops-application using the yaml= applications.yml
? Enter a valid AgentIdentifier: debacligitops
```
Following output:
```
Creating GitOps-Application with id:  gitops-application
Successfully created GitOps-Application with id= gitops-application

```
Updating the existing application:

Same sample YAML, but modified the service name:
```
gitops:
  name: gitops-application
  projectIdentifier: default_project
  orgIdentifier: default
  type: application
  application:
    metadata:
      namespace: default
      clusterName: gitopscluster
      labels:
        harness.io/serviceRef: gitopsservicemodified
        harness.io/envRef: gitopsenv
    spec:
      source:
        repoURL: https://github.com/harness-community/harnesscd-example-apps
        path: guestbook
        targetRevision: master
      destination:
        server: https://kubernetes.default.svc
        namespace: default
  agentIdentifier: AGENT_NAME
  clusterIdentifier: gitopscluster
  repoIdentifier: gitopsrepo
 ```
 Command:
 ```
 harness gitops-application --file applications.yml apply
 ```
Output:

```
Trying to create or update gitops-application using the yaml= application.yml
Found GitOps-Application with id= gitops-application
Updating details of GitOps-Application with id= gitops-application
Syncing the Gitops-Application: gitops-application
Successfully synced GitOps-Application with id= gitops-application
Successfully updated repository with id= gitops-application
```

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -61,6 +61,7 @@ func Put(reqUrl string, auth string, body interface{}, contentType string, reque
 
 func Get(reqUrl string, auth string) (respBodyObj ResponseBody, err error) {
 	req, err := http.NewRequest("GET", reqUrl, nil)
+	//fmt.Println("req:", req)
```
Comment (devinder-harness) [inline client.go:64]: remove commented line

File: gitops-application.go [hunk 2]
```diff
@@ -38,15 +38,24 @@ func applyGitopsApplications(c *cli.Context) error {
 		"clusterIdentifier": clusterIdentifier,
 		"repoIdentifier":    repoIdentifier,
 	})
+
+	applicationName := valueToString(GetNestedValue(requestBody, "gitops", "name").(string))
+	syncApplicationURL := GetUrlWithQueryParams("", baseURL,
+		fmt.Sprintf(GITOPS_APPLICATION_ENDPOINT+"/%s", applicationName+"/sync"), map[string]string{
```
Comment (devinder-harness) [inline gitops-application.go:44]: move entire string to constant and pass variable:
```
const GITOPS_APPLICATION_ENDPOINT = "applications/%s/sync"
 fmt.Sprintf(GITOPS_APPLICATION_ENDPOINT, applicationName)
 ```

File: gitops-application.go [hunk 3]
```diff
@@ -38,15 +38,24 @@ func applyGitopsApplications(c *cli.Context) error {
...
+			"projectIdentifier": projectIdentifier,
+		})
+
 	extraParams := map[string]string{
 		"agentIdentifier": agentIdentifier,
 	}
-	applicationName := valueToString(GetNestedValue(requestBody, "gitops", "name").(string))
-
-	entityExists := getEntity(baseURL, fmt.Sprintf(GITOPS_APPLICATION_ENDPOINT+"/%s"+"/exists", applicationName),
+	entityExists := getEntity(baseURL, fmt.Sprintf(GITOPS_APPLICATION_ENDPOINT+"/%s", applicationName),
```
Comment (devinder-harness) [inline gitops-application.go:54]: same as above

File: gitops-application.go [hunk 4]
```diff
@@ -109,26 +127,28 @@ func createGitOpsApplicationPayload(requestBody map[string]interface{}) GitOpsAp
...
-			},
-			Namespace: valueToString(GetNestedValue(requestBody, "gitops", "application", "metadata", "namespace").(string)),
-		},
-		Spec: Spec{
-			Source: Source{
-				RepoURL:        valueToString(GetNestedValue(requestBody, "gitops", "application", "spec", "source", "repoURL").(string)),
-				Path:           valueToString(GetNestedValue(requestBody, "gitops", "application", "spec", "source", "path").(string)),
-				TargetRevision: valueToString(GetNestedValue(requestBody, "gitops", "application", "spec", "source", "targetRevision").(string)),
+func createGitOpsApplicationPUTPayload(requestBody map[string]interface{}) GitOpsApplication {
+	putApp := GitOpsApplication{
```
Comment (devinder-harness) [inline gitops-application.go:131]: nit: We can change the variable name to improve readability. Ex: `GitOpsApplicationData` / `GitopsAppDetails` (these are just examples, feel free to rename whatever is readable acc to you)

File: gitops-application.go [hunk 5]
```diff
@@ -56,6 +65,7 @@ func applyGitopsApplications(c *cli.Context) error {
 				getColoredText(applicationName, color.FgBlue))
 			return nil
 		}
+
```
Comment (Ompragash) [inline gitops-application.go:68]:
```suggestion
```

File: gitops-application.go [hunk 6]
```diff
@@ -70,10 +80,18 @@ func applyGitopsApplications(c *cli.Context) error {
 				"clusterIdentifier": clusterIdentifier,
 			})
 		newAppPayload := createGitOpsApplicationPUTPayload(requestBody)
-		_, err = Put(appPUTUrl, cliCdRequestData.AuthToken, newAppPayload, CONTENT_TYPE_JSON, nil)
+		syncPayload := createGitOpsApplicationPayload(requestBody)
+		println("Syncing the Gitops-Application:", getColoredText(applicationName, color.FgGreen))
```
Comment (Ompragash) [inline gitops-application.go:84]:
```suggestion
		println("Syncing the GitOps app before updating the spec:", getColoredText(applicationName, color.FgGreen))
```

File: gitops-application.go [hunk 7]
```diff
@@ -70,10 +80,18 @@ func applyGitopsApplications(c *cli.Context) error {
...
+		_, err = Post(syncApplicationURL, cliCdRequestData.AuthToken, syncPayload, CONTENT_TYPE_JSON, nil)
 		if err == nil {
-			println(getColoredText("Successfully updated repository with id= ", color.FgGreen) +
+			println(getColoredText("Successfully synced GitOps-Application with id= ", color.FgGreen) +
```
Comment (Ompragash) [inline gitops-application.go:87]:
```suggestion
			println(getColoredText("Successfully synced GitOps app with id= ", color.FgGreen) +
```

File: gitops-application.go [hunk 8]
```diff
@@ -70,10 +80,18 @@ func applyGitopsApplications(c *cli.Context) error {
...
 				getColoredText(applicationName, color.FgBlue))
+			_, err = Put(appPUTUrl, cliCdRequestData.AuthToken, newAppPayload, CONTENT_TYPE_JSON, nil)
+			if err == nil {
+				println(getColoredText("Successfully updated repository with id= ", color.FgGreen) +
```
Comment (Ompragash) [inline gitops-application.go:91]:
```suggestion
				println(getColoredText("Successfully updated GitOps app with id= ", color.FgGreen) +
```

File: gitops-application.go (hunk 4 above)
Comment (Ompragash) [reply gitops-application.go:131]: +1

File: gitops-application.go (hunk 2 above)
Comment (Ompragash) [reply gitops-application.go:44]: `/sync` endpoint comes in only for the Sync operation and `GITOPS_APPLICATION_ENDPOINT` is also used in three other places so it's safe to skip from the constant.

Comment (Ompragash) [inline gitops-application.go:44]:
```suggestion
		fmt.Sprintf(GITOPS_APPLICATION_ENDPOINT + "/%s", applicationName + "/sync"), map[string]string{
```

File: gitops-application.go (hunk 3 above)
Comment (Ompragash) [reply gitops-application.go:54]: -1

Comment (Ompragash) [inline gitops-application.go:54]:
```suggestion
	entityExists := getEntity(baseURL, fmt.Sprintf(GITOPS_APPLICATION_ENDPOINT + "/%s", applicationName),
```

File: gitops-application.go [hunk 9]
```diff
@@ -56,6 +65,7 @@ func applyGitopsApplications(c *cli.Context) error {
 				getColoredText(applicationName, color.FgBlue))
 			return nil
 		}
+
 	} else {
 		println("Found GitOps-Application with id=", getColoredText(applicationName, color.FgCyan))
```
Comment (Ompragash) [inline gitops-application.go:70]:
```suggestion
		println("Found GitOps Application with id=", getColoredText(applicationName, color.FgCyan))
```

File: gitops-application.go [hunk 10]
```diff
@@ -56,6 +65,7 @@ func applyGitopsApplications(c *cli.Context) error {
...
 		println("Updating details of GitOps-Application with id=", getColoredText(applicationName, color.FgBlue))
```
Comment (Ompragash) [inline gitops-application.go:71]:
```suggestion
		println("Updating details of GitOps Application with id=", getColoredText(applicationName, color.FgBlue))
```

File: client.go (hunk 1 above)
Comment (Ompragash) [reply client.go:64]: +1

File: gitops-application.go (hunk 2 above)
Comment (neelam-harness) [reply gitops-application.go]: Let's declare the `sync` in the same way as `GITOPS_APPLICATION_ENDPOINT` and use it here like:
`fmt.Sprintf(GITOPS_APPLICATION_ENDPOINT+"/%s", applicationName+GITOPS_SYNC_ENDPOINT)`

Adding it in the constants also gives reader an opportunity to see that this endpoint is being used somewhere in GitOps flow which goes missing if we just hard-code it.

Comment (Ompragash) [reply gitops-application.go]: IMO we use CONSTANTS as an easy way around VARIABLES (i.e. for things that can be changed/modified as per user needs) and adding more constants that too for `/sync` only increases the list.

Maybe we can use code comments to help the contributors of the future.

**Review by Ompragash — APPROVED** [review]

**Pull request merged**